### PR TITLE
Adding whitelist methods needed for DefaultGroovyMethods

### PR DESCRIPTION
**Pull request by abayer**
Might be better organized, might be things here that are problems. `invokeMethod` was needed for `sum`, so I've removed `sum` from the tests I'll be pushing up over in `workflow-cps` in an hour or so.

cc @reviewbybees

### CODE REVIEW

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist [hunk 1]
```diff
@@ -1,6 +1,169 @@
...
+method org.codehaus.groovy.runtime.callsite.BooleanClosureWrapper callForMap java.util.Map$Entry
+new groovy.util.PermutationGenerator java.util.Collection
+new java.util.ArrayList
+new java.util.ArrayList int
+new java.util.LinkedHashMap
+new org.codehaus.groovy.runtime.ReverseListIterator java.util.List
+new org.codehaus.groovy.runtime.callsite.BooleanClosureWrapper groovy.lang.Closure
+new org.codehaus.groovy.util.ArrayIterator java.lang.Object[]
+staticField groovy.lang.Closure DONE
+staticMethod com.cloudbees.groovy.cps.CpsDefaultGroovyMethods $addEntry__java_util_Map__java_lang_Object java.util.Map java.lang.Object
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: You should not need any of these.

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist [hunk 2]
```diff
@@ -1,6 +1,169 @@
...
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods tail java.util.List
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods toList java.lang.Object[]
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods toList java.lang.String
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods withDefault java.util.List groovy.lang.Closure
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods withDefault java.util.Map groovy.lang.Closure
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods withLazyDefault java.util.List groovy.lang.Closure
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethodsSupport createSimilarCollection java.util.Collection
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethodsSupport createSimilarCollection java.util.Collection int
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethodsSupport createSimilarMap java.util.Map
+staticMethod org.codehaus.groovy.runtime.InvokerHelper getMetaClass java.lang.Object
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: Sounds dangerous, perhaps. What needs this?

**Comment by jglick** [comment]
FTR this is a follow-up to https://github.com/jenkinsci/workflow-cps-plugin/pull/124.

**Comment by abayer** [comment]
Will fix `sanity`. And am trying with the others removed now.

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist (hunk 2 above)
Comment (abayer) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: Nothing, it looks like, but still checking.

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist (hunk 1 above)
Comment (abayer) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: Turns out we do! Don't know why.

Comment (abayer) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: Looks like these are all private methods, fwiw.

**Comment by abayer** [comment]
@jglick Ok, so the private methods are still needed but otherwise I think we're clear here.

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist [hunk 3]
```diff
@@ -1,4 +1,13 @@
-# groovy-cps
+# groovy-cps - not needed from workflow-cps 2.33+
+staticMethod com.cloudbees.groovy.cps.CpsDefaultGroovyMethods $addEntry__java_util_Map__java_lang_Object java.util.Map java.lang.Object
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: Why are we _adding_ to this list, if 2.32- had only `each`, and 2.33+ looks at the natural method signature? I am not following.

Comment (abayer) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: I'll try removing that one.

**Comment by abayer** [comment]
@jglick What's special about the private methods is that they have to be white listed in their `CpsDefaultGroovyMethods` forms. Otherwise it doesn't work.

**Comment by jglick** [comment]
Right, got that, but the question is why. Perhaps the `CpsWhitelist` code needs to scan `getDeclaredMethods` rather than just `getMethods`?

**Comment by abayer** [comment]
@jglick That sounds right, yeah. I'll try it.

**Comment by abayer** [comment]
Hrm, nope, `getDeclaredMethods()` makes everything worse. Experimenting further. =)

**Comment by abayer** [comment]
@jglick Ok, https://github.com/jenkinsci/workflow-cps-plugin/pull/139 now updated to automatically pass for private methods in the category. That seems the best approach I can come up with.

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist [hunk 4]
```diff
@@ -213,16 +262,34 @@ staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getAt java.util.Li
 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getAt java.util.Map java.lang.Object
 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getAt java.util.regex.Matcher int
 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getChars java.lang.String
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods groupBy java.util.Collection groovy.lang.Closure
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods groupBy java.util.Collection java.lang.Object[]
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods groupBy java.util.Map groovy.lang.Closure
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods groupBy java.util.Map java.lang.Object[]
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods groupEntriesBy java.util.Map groovy.lang.Closure
+staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods inject java.lang.Object groovy.lang.Closure
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: Reminder to self: this is not at all as scary as it sounds! More or less `foldl` or similar.

Speaking of which, fun party trick: http://foldl.com/ (alas, seems to be dead now)

**Comment by abayer** [comment]
@jglick We need a release of `script-security` now, I think.

**Pull request merged**